Find regions with abnormally large inserts.

danvk opened the pull request:
This is pretty simple and will conflict with @arahuja's #325, but it is at least able to find interesting regions  in synth4.

For example, here's one of the output lines when I run it on synth4 tumor:

    GenomeRange(21,10744200,10760575)

and the corresponding view in IGV:

![large gaps](https://cloud.githubusercontent.com/assets/98301/8833677/1615b138-307e-11e5-8131-fd58cec3e95a.png)

it's sorted by insert size, so you're just seeing the big gaps. But there does seem to be some evidence for a large, consistent deletion.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/guacamole/334)
<!-- Reviewable:end -->

danvk commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+    override val description = "Find structural variants, i.e. large insertions or deletions, inversions and translocations."
+
+    def isFirstInPair(read: MappedRead) = {
+      read.matePropertiesOpt match {
```
Is this a reasonable way to get at `isFirstInPair`, or is there a simpler way to do it with `Option` types?

arahuja replied:
I think using the collection functions is preferred over matching on `Some`/`None` 

```
read.matePropertiesOpt.exists(_.isPaired)
```

danvk commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (insertSize > 0) && (insertSize < MAX_INSERT_SIZE)
+      })
+      readsInRange.persist()
```
Is there a rule of thumb about which RDDs you're supposed to `.persist()`?

arahuja replied:
If you are going to use the RDD twice, it might make sense to persist.  I rarely do as most are constructed once and then either output or shuffled, both operations which make no use of the persistance

danvk replied:
I use `readsInRange` on both lines 96 (for getting stats on insert sizes) and on line 102 (for building the per-block list of insert sizes).

If I didn't have the `.persist()` here, would it have to re-run the `mappedPairedReads` and `.filter(_.isFirstInPair)` steps?

danvk commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      println("Stats on inferredInsertSize: " + insertStats)
+
+      // Create a map from (block) -> [insert sizes]
+      // TODO: take greater advantage of data locality than .groupByKey() does?
```
I'm curious to hear your thoughts on this. The idea is that for an input which is sorted by position, nearby records will have similar blocks. So a combiner that operates on individual partitions would be very effective. I'm not sure if `.groupByKey` does that.

arahuja replied:
You may be looking for something like `combineByKey` then which you can two functions, a per partition combiner and then a per-key combiner. A `Combiner` and `Reducer` from the map-reduce world.

If possible, those are usually better to use anyways, since `groupByKey` pulls all the values of a key into memory.

You could also use `repartitionAndSortWithinPartitions` if you want your values in sorted order when processing.  

Haven't looked through what this is doing yet though to see if this applies.

danvk replied:
I think I may want `aggregateByKey`, which lets you specify a reduce function, so that all the values don't have to go into memory at once.

ryan-williams replied:
Yea, I think either one will work; if I have it right, `combineByKey` is more general, but for this case all you need is `aggregateByKey`.

You are going from an `RDD[(K, V)]` to an `RDD[(K, U)]`. You have:
* a `zeroValue: U`,
* a `seqOp: (U, V) => U`, and
* a `combOp: (U, U) => U`,

which are the necessities for the `aggregateByKey` API.

`combineByKey` generalizes further by substituting a `createCombiner: V => U` instead of `zeroValue: U`.

`K == Location`, `V == Long`, `U == StatsCounter`, `zeroValue == new StatCounter()`, and `seqOp` and `combOp` are overloaded versions of `StatsCounter.merge` that take a `Double` and a `StatsCounter`, resp.

ryan-williams replied:
```
      val exceptionalBlocks: RDD[Location] =
        (for {
          r <- readsInRange
          insertSize <- orientedInsertSize(r)
          pos <- matedReadBlocks(r)
        } yield {
          Location(r.read.referenceContig, pos) -> insertSize
        }).aggregateByKey(
          new StatCounter()
        )(
          seqOp = _ merge _,
          combOp = _ merge _
        )
        .filter(_._2.mean > exceptionalThreshold)
        .map(_._1)
```

arahuja commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,129 @@
+object StructuralVariant {
+
+  val MAX_INSERT_SIZE = 25000
+  val BLOCK_SIZE = 25
```
Can you define/explain a 'block' somewhere. I read this initially thinking it had something to do with the BAM file.

danvk replied:
Sure, I added a clarifying comment in the scaladoc for the caller.

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      position: Long) {
+  }
+
+  case class GenomeRange(
```
Just noting that equivalents of these two case classes seem to proliferate around genomics codebases.

Using BDG [`Feature`](https://github.com/bigdatagenomics/bdg-formats/blob/de290eb5e5e237e13ee20ca9b162744888721041/src/main/resources/avro/bdg.avdl#L710)s / [`ReferenceRegion`](https://github.com/bigdatagenomics/adam/blob/a68a7e1ad3d970111d007b6a4011365e15e213a6/adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala#L125)s would give you writing-to-Parquet for free, instead of a text file, fwiw.

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+
+  object Caller extends SparkCommand[Arguments] {
+    override val name = "structural-variant"
+    override val description = "Find structural variants, i.e. large insertions or deletions, inversions and translocations."
```
it doesn't do all of this yet, does it?

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    // Coalesce sequences of values separated by blockSize into single ranges.
+    // The sequence of values must be sorted.
```
A weird apparent pre-condition of this is also that no values in `xs` are separated by less than `blockSize`?

timodonnell replied:
A bit more detail on what the input and especially output are of this function would be helpful.

ryan-williams replied:
Also probably worth commenting that the ranges returned are inclusive pairs of "block start" loci, if I have that right?

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+        Seq[Long]()
+      } else {
+        val mateStart = r.mateAlignmentProperties.map(mp => mp.start).getOrElse(r.read.start)
+        val start = List(r.read.start, r.read.end, mateStart).min
```
can `r.read.end` be less than `r.read.start`?

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      } else {
+        val mateStart = r.mateAlignmentProperties.map(mp => mp.start).getOrElse(r.read.start)
+        val start = List(r.read.start, r.read.end, mateStart).min
+        val roundedStart = (1.0 * start / BLOCK_SIZE).toLong * BLOCK_SIZE
```
why not just `start / BLOCK_SIZE * BLOCK_SIZE`? integer division ftw…

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+        val start = List(r.read.start, r.read.end, mateStart).min
+        val roundedStart = (1.0 * start / BLOCK_SIZE).toLong * BLOCK_SIZE
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (roundedStart to (roundedStart + insertSize) by BLOCK_SIZE)
```
intellij gives me an "unnecessary parends" warning here and on L117, if we care about those

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      val firstInPair = pairedMappedReads.filter(_.isFirstInPair)
+
+      // Pare down to mate pairs which aren't highly displaced & aren't inverted.
+      val readsInRange = firstInPair.filter(r => {
```
perhaps:
```
firstInPair.filter(r => orientedInsertSize(r).exists(_ < MAX_INSERT_SIZE))
```
or
```
firstInPair.filter(r => orientedInsertSize(r).exists(s => 0 < s && s < MAX_INSERT_SIZE))
```
depending whether you actually care about the `== 0` case or were just using `0` as a sentinel…

ryan-williams replied:
more golfed:

```
firstInPair.flatMap(orientedInserSize).filter(_ < MAX_INSERT_SIZE)
```

ryan-williams replied:
and while we're here, I increasingly find `for`-comps the most readable
```
for {
  r <- firstInPair
  s <- orientedInsertSize(r)
  if 0 < s && s < MAX_INSERT_SIZE
} yield s
```

ryan-williams replied:
`insertSizes.filter(_ < MAX_INSERT_SIZE)`

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      })
+      readsInRange.persist()
+
+      val insertSizes = readsInRange.flatMap(orientedInsertSize)
```
this {c,sh}ould be declared above L94; it would make that block cleaner (my 3rd comment there used this)

ryan-williams replied:
ah nvm, my mistake.

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      // Create a map from (block) -> [insert sizes]
+      // TODO: take greater advantage of data locality than .groupByKey() does?
+      val insertsByBlock = readsInRange.flatMap(r => {
+        val insertSize = orientedInsertSize(r).getOrElse(0)
```
again, maybe want to just start out with `insertSizes` here?

ryan-williams replied:
```
(for {
  r <- readsInRange
  pos <- matedReadBlocks(r)
  insertSize <- orientedInsertSize(r)
} yield {
  Location(r.read.referenceContig, pos) -> insertSize
}).groupByKey
```

Tricky thing about `for`s and chained `flatMap`s and `map`s in Scala: you can't `flatMap` over an `Option` then a list; e.g. this would give a not-very-descriptive error about inferred type something somethings:
```
  r <- readsInRange
  insertSize <- orientedInsertSize(r)
  pos <- matedReadBlocks(r)
```

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      rangeStarts += xs.length
+
+      rangeStarts.sliding(2).map(items => {
+        val startIdx = items(0)
```
intellij suggests `items.head` fwiw

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      }
+      rangeStarts += xs.length
+
+      rangeStarts.sliding(2).map(items => {
```
ICYMI:
```
for { 
  (startIdx, nextStartIdx) <- rangeStarts.sliding(2) 
} yield { 
  (xs(startIdx), xs(nextStartIdx - 1)) 
}
```

ryan-williams replied:
good work using `sliding` though, btw :) I was looking for a way to suggest you use it on L62

ryan-williams replied:
another syntax that's useful sometimes:
```
rangeStarts.sliding(2).map {
  case (startIdx, nextStartIdx) => (xs(startIdx), xs(nextStartIdx - 1))
}
```

ryan-williams replied:
OK, in today's edition of "you can always do it in one pass using `foldLeft`", I think this works:

```
    def coalesceAdjacent(xs: Array[Long], blockSize: Long): Iterable[(Long, Long)] = {
      xs match {
        case Array(first, rest @ _*) =>
          val (list, last) =
            rest.foldLeft(
              (ListBuffer[(Long, Long)](), (first, first))
            )(
                (p, pos) => {
                  val (arr, (curStart, curEnd)) = p
                  if (curEnd + blockSize == pos)
                    (arr, (curStart, pos))
                  else
                    (arr :+(curStart, curEnd), (pos, pos))
                }
              )
          list :+ last
        case _ => Nil
      }
    }
```

Not claiming this is cleaner or better necessarily, but worth a look :)

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+
+    // Coalesce sequences of values separated by blockSize into single ranges.
+    // The sequence of values must be sorted.
+    def coalesceAdjacent(xs: IndexedSeq[Long], blockSize: Long): Iterator[(Long, Long)] = {
```
`IndexedSeq[Long]` seems like an unusual type to see here; any reason for it? Seems like you only pass an `Array[Long]` here

ryan-williams commented on src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala:
```diff
@@ -0,0 +1,133 @@
+      })
+
+      // Group adjacent blocks on the same contig
+      val exceptionalRanges = exceptionalBlocks.groupBy(_._1.contig).flatMap(kv => {
```
I am guessing this will work for you with the size the array you anticipate dealing with, but the `groupBy`, plus the `.toArray.sorted` below, feels painful.

I *think* we can use some `{aggregate,combine}ByKey` trickery like I gave an example of above, as well as some thinking along the lines of my `foldLeft` example even further above, and do all of this in `RDD`-space without having to pull the whole values list into memory.

I'll see if I can get an example together for you.

ryan-williams replied:
hm, yea not coming up with much... to do anything smarter you'd have to go into full "Region Join" territory, I think. Anyways you have at most a few million `Long`s in an array here, so it should be OK.

ryan-williams commented on src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala:
```diff
@@ -0,0 +1,16 @@
+      150L)
+    val ranges = StructuralVariant.Caller.coalesceAdjacent(nums, 10)
+    assert(ranges.toList === List((0L, 30L), (50L, 50L), (70L, 80L), (100L, 120L), (150L, 150L)))
+  }
```
a few more test cases would be nice. One I did in my shell was:

```
scala> coalesceAdjacent(Array(25,50,75,101,126,200,250,275), 0)
res2: Iterable[(Long, Long)] = ListBuffer((25,75), (101,126), (200,200), (250,275))
```

ryan-williams replied:
also some tests of the calling would be good

danvk commented:
Thanks for all the feedback Ryan, Arun & Tim. I believe I've addressed all the comments, please take another look!

---

Review status: 0 of 3 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 41 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-41)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L41)):</sup>
But mine are subtly different :)

Point taken. Both those types look like they're for regions, though, rather than individual loci?

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 48 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-48)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L48)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 49 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r1-48)** ([raw file](https://github.com/hammerlab/guacamole/blob/5949189588c7e80069019ac15f1a25567a1998ea/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L49)):</sup>
(I deleted this code after rebasing on Arun's PairedRead refactor)

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 58 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-58)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L58)):</sup>
I expanded the comment. Let me know if you guys think there's a clearer way of structuring this; code whose explanation is as long as the code is usually no good.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 59 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-59)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L59)):</sup>
`IndexedSeq[Long]` is the more general type, right?

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 68 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-68)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L68)):</sup>
I went with the for comprehension version. Had to make it `Seq(startIdx, nextStartIdx)` though.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 69 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-69)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L69)):</sup>
(this is obsolete now that I'm using the for comprehension)

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 81 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-81)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L81)):</sup>
Signs point to no. I switched this to use `Math.min` on the two possibilities.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 82 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-82)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L82)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 84 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-84)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L84)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 94 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-94)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L94)):</sup>
I do need to filter for the case where `orientedInsertSize < 0`. It corresponds to an inversion. 

I like the for comprehension version.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 105 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r1-105)** ([raw file](https://github.com/hammerlab/guacamole/blob/5949189588c7e80069019ac15f1a25567a1998ea/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L105)):</sup>
I used the for comprehension mostly as you suggested. Thanks for the code!

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 121 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-121)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L121)):</sup>
I rewrote this use `case` parameter matching. Could it be done more cleanly with another for comprehension?

---

<sup>**[src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala, line 15 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7_--r3-15)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala#L15)):</sup>
Hmm, I wouldn't consider `blockSize` of zero and values which aren't multiples of `blockSize` to be valid inputs.

I'd like to add some tests. Do you have advice about how to craft BAM files from scratch? Just type out lines of a SAM file and convert?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/334)
<!-- Sent from Reviewable.io -->

danvk commented:
Friendly ping

ryan-williams commented:
Looking now, sorry for the delay

ryan-williams commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 41 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-41)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L41)):</sup>
The exact equivalents would be ADAM's [ReferencePosition](https://github.com/bigdatagenomics/adam/blob/adam-parent_2.10-0.17.0/adam-core/src/main/scala/org/bdgenomics/adam/models/ReferencePosition.scala#L86) and [ReferenceRegion](https://github.com/bigdatagenomics/adam/blob/adam-parent_2.10-0.17.0/adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala#L125).


Any comment on the point about writing Features to Parquet instead of bespoke text file? "Sounds like a maybe a good idea later" is a fine response, j/w.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 57 \[r4\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jv-dGykX0gmviyOmDKj)** ([raw file](https://github.com/hammerlab/guacamole/blob/d07e54906ac51da590e4eea75b3d5cfc31d46b2d/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L57)):</sup>
Still a little confusing imho, but not a big deal. One attempt at rephrasing fwiw:

Given a sorted sequence of positions (all of which are multiples of @blockSize), coalesce ranges of consecutive multiples of @blockSize into tuples representing the start and end (inclusive) of each range.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 81 \[r4\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jv-f6xo_GfRiUQzwyI0)** ([raw file](https://github.com/hammerlab/guacamole/blob/d07e54906ac51da590e4eea75b3d5cfc31d46b2d/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L81)):</sup>
`s/mp => mp/_/`

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 115 \[r4\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jv-gIYL-7eNLUWcMPh9)** ([raw file](https://github.com/hammerlab/guacamole/blob/d07e54906ac51da590e4eea75b3d5cfc31d46b2d/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L115)):</sup>
heh, didn't want to go for `(seqOp = _ merge _, combOp = _ merge _)`? this lgtm too, j/w

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 120 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-121)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L120)):</sup>
Sure. My earlier comment was about trying to avoid the `.toArray` below, but in response to your q:
```
val exceptionalRanges =
  for {
    (contig, locations) <- exceptionalBlocks.groupBy(_.contig)
    positions = locations.map(_.position).toArray.sorted
    (start, end) <- coalesceAdjacent(positions, BLOCK_SIZE)
  } yield {
    GenomeRange(contig, start, end)
  }
```

As a rule, any time you have a `map` inside one or more `flatMap`s, you can make a `for` out of it; you don't have to, jfyi.

Also this rule is good enough that I remember it and not even why it's true without thinking about it for a minute :)

---

<sup>**[src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala, line 15 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7_--r3-15)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala#L15)):</sup>
Oh, sorry, I didn't mean that `0` for `blockSize` above; I had hard-coded `25` in the impl of `coalesceAdjacent` in my repl, so that `0` was just a placeholder. With that example I just meant to show you a more complicated test of the functionality. I see now that it breaks the precondition of everything being a multiple of `blockSize`; feel free to disregard.

re: calling tests... yea we don't have a great story there. Every time I'm in this position I put N units of work into improving that story and generate issues related to it worth 2N units of work.

Open to using this to try to improve the story there, but also open to punting to a later PR.

One answer is make a SAM file with many copies of one read, then manually change each one's start/end and mate info to create the test you want, and remove as many extraneous fields as you can.

Another (probably better) answer is to just construct your reads (with mate info) manually in scala in your test case.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/334)
<!-- Sent from Reviewable.io -->

danvk commented:
Addressed all comments, please take another look.

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 41 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-41)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L41)):</sup>
Eventually it should output [`StructuralVariant`][1] records, but this is all pretty in flux right now.

[1]: https://github.com/bigdatagenomics/bdg-formats/blob/de290eb5e5e237e13ee20ca9b162744888721041/src/main/resources/avro/bdg.avdl#L373

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 57 \[r4\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jv-dGykX0gmviyOmDKj)** ([raw file](https://github.com/hammerlab/guacamole/blob/d07e54906ac51da590e4eea75b3d5cfc31d46b2d/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L57)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 58 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-58)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L58)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 81 \[r4\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jv-f6xo_GfRiUQzwyI0)** ([raw file](https://github.com/hammerlab/guacamole/blob/d07e54906ac51da590e4eea75b3d5cfc31d46b2d/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L81)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 115 \[r4\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jv-gIYL-7eNLUWcMPh9)** ([raw file](https://github.com/hammerlab/guacamole/blob/d07e54906ac51da590e4eea75b3d5cfc31d46b2d/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L115)):</sup>
Yeah, paren-less function calling scares me a little bit.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 120 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-121)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L120)):</sup>
Cool. The destructuring assignment seems like a big win of using `for` comprehensions.

---

<sup>**[src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala, line 15 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7_--r3-15)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala#L15)):</sup>
Pure end-to-end tests of MapReduces tend to be really brittle. Any little change you make will change the output in complex ways.

The best way to test pipelines like this is to break them down into isolated bits that have more well-defined behavior, e.g. "load the mated, mapped reads with insert sizes in a given range", "find the median insert size for a collection of reads" and "group these locations into contiguous ranges" and test each of those independently.

I'd like to do that once this code settles down a bit (it's going to change dramatically as I implement something more DELLY-like).

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/334)
<!-- Sent from Reviewable.io -->

arahuja commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala, line 120 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7Zz-r3-121)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala#L120)):</sup>
Sorry for the late add here, and this can probably be handled later, but the positions can be sorted without loading them into an array first, but it's a little annoying

i.e. 
```
class ContigPartitioner(tasks: Int) extends Partitioner {
   def partition(key: Any) {
        case key: Location(contig, start, end) => contig //or something to map it to an int
    }
}


exceptionalBlocks.keyBy(somethingThatIsLocation).repartitionAndSort(new ContigPartitioner())

exceptionalBlocks.mapPartitions( things => {
  while (things)
       while (notCoalesced) {
          keepCoalescing
       }
       yield coalesced
   }
 } 
```
etc ... written in Reviewable/not Intellij  and Location would need an ordering

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/334)
<!-- Sent from Reviewable.io -->

ryan-williams commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

<sup>**[src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala, line 15 \[r3\]](https://reviewable.io:443/reviews/hammerlab/guacamole/334#-Jur141mDoaPaQaHo7_--r3-15)** ([raw file](https://github.com/hammerlab/guacamole/blob/86aec297b422922dae2c9c224b1e25ad7304d4f2/src/test/scala/org/hammerlab/guacamole/commands/StructuralVariantCallerSuite.scala#L15)):</sup>
I'm saying that you can (and we do typically when we test stuff like this) create a dummy `RDD[GenomeRange]` or `RDD[MappedRead]` in scala pretty easily.

```
sc.parallelize(List(MappedRead(…), MappedRead(…), MappedRead(…)))
```

Then you can write some very simple tests of just the logic you're adding here, without worrying about I/O.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/334)
<!-- Sent from Reviewable.io -->
